opm/parser/eclipse/EclipseState/Schedule/Compsegs.cpp

```diff
@@ -0,0 +1,136 @@
+
+        // only handle the second record here
+        // The first record here only contains the well name
+        std::vector<std::shared_ptr<Compsegs>> compsegs;
+
+        for (size_t recordIndex = 1; recordIndex < compsegsKeyword->size(); ++recordIndex) {
+            DeckRecordConstPtr record = compsegsKeyword->getRecord(recordIndex);
+            // following the coordinate rule for completions
+            const int I = record->getItem("I")->getInt(0) - 1;
+            const int J = record->getItem("J")->getInt(0) - 1;
+            const int K = record->getItem("K")->getInt(0) - 1;
+            const int branch = record->getItem("BRANCH")->getInt(0);
+
+            double distance_start;
+            double distance_end;
+            if (record->getItem("DISTANCE_START")->hasValue(0)) {
+                distance_start = record->getItem("DISTANCE_START")->getRawDouble(0);
+            } else if (recordIndex == 1) {
+                distance_start = 0.;
+            } else {
+                // TODO: the end of the previous connection or range
+                // 'previous' should be in term of the input order
+                // since basically no specific order for the completions
+                throw std::runtime_error("this way to obtain DISTANCE_START not implemented yet!");
+            }
+            if (record->getItem("DISTANCE_END")->hasValue(0)) {
+                distance_end = record->getItem("DISTANCE_END")->getRawDouble(0);
+            } else {
+                // TODO: the distance_start plus the thickness of the grid block
+                throw std::runtime_error("this way to obtain DISTANCE_END not implemented yet!");
+            }
+
+            WellCompletion::DirectionEnum direction;
+            if (record->getItem("DIRECTION")->hasValue(0)) {
+                direction = WellCompletion::DirectionEnumFromString(record->getItem("DIRECTION")->getString(0));
+            } else if (!record->getItem("DISTANCE_END")->hasValue(0)) {
+                throw std::runtime_error("the direction has to be specified when DISTANCE_END in the record is not specified");
+            }
+
+            int end_IJK;
+            if (record->getItem("END_IJK")->hasValue(0)) {
+                // following the corrdinate rule for completions
+                end_IJK = record->getItem("END_IJK")->getInt(0) - 1;
+                if (!record->getItem("DIRECTION")->hasValue(0)) {
+                    throw std::runtime_error("the direction has to be specified when END_IJK in the record is specified");
+                }
+            } else {
+                // only one completion is specified here
+                end_IJK = -1;
+            }
+
+            double center_depth;
+            if (record->getItem("CENTER_DEPTH")->hasValue(0)) {
+                center_depth = record->getItem("CENTER_DEPTH")->getRawDouble(0);
+            } else {
+                center_depth = 0.;
```

Comment:
OK - but observe that the test `if (record->getItem("CENTER_DEPTH")->hasValue(0))` will **always** return true. The point is that `hasValue( )` does *not* differentiate between the cases "value set in input deck" and "using default value" - they are both valid values; the `hasValue( )` method is mostly relevant when an item does not have a default value. I would instead reccommend:

```C++
// Use Type value for center_depth: https://github.com/OPM/opm-parser/blob/master/opm/parser/eclipse/EclipseState/Util/Value.hpp - to get some support for default handling.

{
     const auto depthItem = record->getItem<ParserKeywords::COMPSEGS::CENTER_DEPTH>(0);
     if (!depthItem->defaultApplied(0))
          center_depth.setValue( depthItem->getSIDouble(0) );
}
```